client/python/conductor/FrinxConductorWrapper.py

```diff
@@ -20,144 +24,209 @@
+        self.task_source = TaskSource()
+
+        self.polling_interval = polling_interval
+
+        wfcMgr = WFClientMgr(server_url, headers=headers)
+        self.taskClient = wfcMgr.taskClient
+        self.worker_id = worker_id or hostname
+
+    def start_workers(self):
+        for i in range(self.consumer_worker_count):
+            thread = Thread(target=self.consume_task)
+            thread.daemon = True
+            thread.start()
+
+        logger.info("Starting a queue polling")
+        fail_count = 0
         while True:
             try:
                 time.sleep(float(self.polling_interval))
-                self.lock.acquire()
-                queuesTemp = self.taskClient.getTasksInQueue("all")
-                log.debug('Queues polled: %s', queuesTemp)
-                self.queues = queuesTemp
-                failCount = 0
-            except Exception as err:
-                log.exception('Unable to read queue info. Error count: %s', failCount, err)
-                self.queues = {}
-                failCount =+ 1
-                if (failCount > 10):
-                    log.exception('Exiting, unable to read queue info')
+                queues_temp = self.taskClient.getTasksInQueue("all")
+                self.task_source.handle_tasks(queues_temp)
+                fail_count = 0
+            except Exception:
+                logger.error(
+                    f"Unable to read a queue info after {fail_count} attempts", exc_info=True
+                )
+                self.task_source.handle_tasks({})
+                fail_count = +1
+                if fail_count > 10:
                     exit(1)
-            finally:
-                self.lock.release()
 
-    # register task metadata into conductor
-    def register(self, task_type, task_definition=None):
-        if task_definition is None:
-            task_definition = DEFAULT_TASK_DEFINITION
-
-        log.debug('Registering task ' + task_type + ' : ' + str(task_definition))
-        task_meta = copy.deepcopy(task_definition)
-        task_meta["name"] = task_type
-        try:
-            r = requests.post(self.conductor_task_url,
-                              data=json.dumps([task_meta]),
-                              headers=self.headers)
-            # response_code = r.status_code
-        except Exception as err:
-            log.exception('Error while registering task ' + traceback.format_exc())
-            raise err
-
-    def poll_and_execute(self, taskType, exec_function, domain=None):
-        poll_wait = 5000
+    def consume_task(self):
+        last_task_type = None
         while True:
-            time.sleep(float(self.polling_interval))
+            next_task: NextWorkerTask = self.task_source.get_next_task(last_task_type)
+            last_task_type = None
 
-            # If there are not tasks indicated in queues, skip actual polling
-            if (not self.tasksInQueue(taskType, domain)):
+            if not next_task:
```

Comment:
shouldn't we do some sleeping here ? doesn't this consume too much CPU ?

Reply:
Yes. CPU usage has dropped significantly.